Docker-less Alpine installation

Chiogros opened the pull request:
Hi,

The docker-based installation is not convenient for everyone.
For example, a docker-less installation would allow LXC/D containers usage, without using tricky tips for using Docker inside of them.

I already successfully installed a standalone Immich.
This branch would be used for scripting a command-line interactive installation.

Thanks for your feedback!

Commit:
Create alpine.sh

Chiogros commented:
There are some tips to run Docker in LXC containers (playing with nesting option if I remember), although IMO it is not the best way to do.

alextran1502 commented:
> There are some tips to run Docker in LXC containers (playing with nesting option if I remember), although IMO it is not the best way to do.

What are the reasons?

Chiogros commented:
Allowing a container to go out of its initial scope, even if sometimes it can be convenient, is not the good way just for running applications.

In the same way as in Docker you could use `--privilegied`, running a LXC container with privilegied rights could lead to some security flaws.
(My) PVE is hosting toons of services, I can't imagine a single one being able to disturb the host system.

It's my point of view, I though it could interest some people. If not, we can close the PR, no problem.

alextran1502 commented:
> Allowing a container to go out of its initial scope, even if sometimes it can be convenient, is not the good way just for running applications.
> 
> In the same way as in Docker you could use `--privilegied`, running a LXC container with privilegied rights could lead to some security flaws. (My) PVE is hosting toons of services, I can't imagine a single one being able to disturb the host system.
> 
> It's my point of view, I though it could interest some people. If not, we can close the PR, no problem.

Fair assessment. I was asking for my knowledge. You are welcome to implement the PR. I bet there are people that don't want to use Docker as a mean of deployment the app

Chiogros commented:
Great news, thanks!

Commit:
Moved docker installation script to installation directory.

Commit:
parse_args(), help message and global variables

-h shows a help message with arguments accepted by the script
--dev executes script with dev=1, portions of code can change their behavior
while looking at dev's value.

Commit:
display_message_box(), shows a message in a box

Used for better looking while script is running.

Commit:
update_repo(), enable community repo and update packages list

Commit:
get_source_code(), download Immich release

Download immich release and extract it in temp directory.
If in dev mode, the source code must be in the temp directory.

Commit:
setup_redis(), install and start on boot redis

Commit:
setup_database(), install and configure PostgreSQL

Configure database with a new password and a new database.
The form ask if we want to use an external database, but it is not implemented.
Some extensions are required

Commit:
setup_server(), build and install server & microservices

Services files for each component allows to start them
on boot.

Commit:
setup_web(), web server

Service file to start on boot

Commit:
setup_proxy(), add new config to Nginx

All existing sites in /etc/nginx/http.d are disabled,
so there are no conflicts for listening on 80 port.

Commit:
setup_machine_learning(), build and install everything for ML

Service file to start ML on boot.
The upload folder is a symlink to the server's upload directory

Commit:
remove_install_files(), remove downloaded code

If not in debug mode, the source code and the archive downloaded
in the temp directory are deleted.
As in debug mode, the source code has been provided, it is not
removed.

Chiogros commented:
There are still a few bugs with Alpine services.
I'm working on.

Commit:
unset NODE_ENV while compiling machine-learning

Commit:
Check for AVX2 flags before installing ML

If flags aren't supported, the ML installation in skipped.

Commit:
Services manage child processes and load env variables.

Shell scripts run node commands, these one cannot be managed by services.
So when receiving a stop command, child processes are grepped and
-kindly- killed.

Also environment variables are needed for running server components.

Commit:
--install to start installation and other help messages

Commit:
Quote database values in env vars

Chiogros commented:
Hey @alextran1502,

I solved last bugs, you can try it yourself. 
If you encounter anything bad or think things different, please tell me, I'll try to do my best.

I tried to implement machine-learning in Alpine.
I better understand why you have not set up machine-learning in alpine...

So for the moment, expected for ML, everything seems to work.
ML installation is skipped.

Thanks for your feedback.

panoti commented:
Hi @Chiogros, your contribution is so interesting. Your scripts are quite readable as well. Would you mind if I give you some questions?
1. Can we put `/web/immich-web`, `/server/immich-server`, `/server/immich-microservices`, `/machine-learning/immich-machine-learning` in `installation` directory like `alpine.sh`? I found that these files may work only with Alpine OS, especially OpenRC.
2. Can we fetch `alpine_ver` directly from current user's OS? As you can see if user try to run your script on another alpine version, it can cause error from package updating.

I also have a plan to create standalone docker image by using PM2. I hope I can re-use your script.

Chiogros commented:
Hi @panoti,

> Can we put /web/immich-web, /server/immich-server, /server/immich-microservices, /machine-learning/immich-machine-learning in installation directory like alpine.sh? I found that these files may work only with Alpine OS, especially OpenRC.

I first thought about the same architecture as you: `./installation/services/openrc/`.
In this way, we could then support systemd-based OS by mkdir-ing `./services/systemd/` and translate openrc services to systemd services.

However, every docker images used are alpine images, excepted for ML. Keeping docker/standalone OS the same would certainly be easier to maintain (same packages name, package manager, shell, ...). But it's a way we could try.

> Can we fetch alpine_ver directly from current user's OS? As you can see if user try to run your script on another alpine version, it can cause error from package updating.

Yes, it's on my todo-list :)

Thanks!

panoti commented:
> 
> I first thought about the same architecture as you: `./installation/services/openrc/`. In this way, we could then support systemd-based OS by mkdir-ing `./services/systemd/` and translate openrc services to systemd services.
>

I really like this organization because it may help us distinguish between service managers. 

>
> However, every docker images used are alpine images, excepted for ML. Keeping docker/standalone OS the same would certainly be easier to maintain (same packages name, package manager, shell, ...). But it's a way we could try.
> 

yeah, but I believe that it will make our source code become mess up if we have more kind of service managers like systemd, sysvinit, ....

> 
> Yes, it's on my todo-list :)
> 

Sound great, thank you.

Limezy commented:
Very interesting PR. I'll use it as a basis for a coming Yunohost package 🚀 !

Chiogros commented:
Hi all,

I tried to find a way to run TensorFlow on Alpine, without success.
I've found a [packaged `glibc` for Alpine](https://github.com/sgerrand/alpine-pkg-glibc), but I have a sticky `deps/lib/libtensorflow.so.2: __memcpy_chk: symbol not found` at runtime.

The only thing we could get for the moment for Alpine is an installation without ML.

However, I'll look for a Debian full installation script. In this way, ML could be installed on the same machine as DB and server.

To be continued...

panoti commented:
Debian/Ubuntu may be more popular than alpine. Machine learning search is optional, so if you wanna continue using alpine, just skip it in default.

Commit:
Some functions' documentation

Commit:
Moved openrc service files to installation/services/openrc directory.

It will allows to develop new service files for systemd in parallel in openrc

ManuLinares commented on installation/alpine.sh:
```diff
@@ -0,0 +1,566 @@
+
+# Immich server
+# Install Immich server and microservices.
+# It asks for informations, builds and installs projet, 
```
projet != project

Chiogros replied:
oops

Commit:
Check OS version before installing.

Alpine version is checked from /etc/alpine-version.

Commit:
"Project" typo

Commit:
Rename os version variable.

$needed_os_version has been renamed to $alpine_ver,
this will avoid to change variable's name in the whole script.

Commit:
Update immich version to 1.27

Commit:
Vim typo: usernamei -> username

panoti commented on installation/alpine.sh:
```diff
@@ -0,0 +1,591 @@
+
+	if ! grep -qE "^$alpine_ver" "$os_ver_file"
+	then
+		echo "Alpine $(cat $os_ver_file) not supported: need Alpine $needed_os_ver."
```
you forgot to rename this var `needed_os_ver`

Chiogros replied:
thanks

panoti commented on installation/services/openrc/immich-machine-learning:
```diff
@@ -0,0 +1,34 @@
+required_dirs="$data_path"
+required_files="$data_path/entrypoint.sh"
+
+depend() {
```
hey @Chiogros, why don't you insert LF in front of `{` like `start_pre()`?

Chiogros replied:
LF?

Chiogros replied:
idk, it's a mistake. thanks

panoti replied:
I mean
```
depend()
{
  ...
}
```

Commit:
Rename a forgotten alpine_ver variable.

Commit:
LF before depend first curly bracket

Commit:
Comments in services files

alextran1502 commented on installation/alpine.sh:
```diff
@@ -0,0 +1,591 @@
+
+# All-in-one installation without Docker
+
+immich_ver=1.27.0_37-dev
```
So this one is hardcoded; I assume we must change this per release?

panoti replied:
@Chiogros FYI, you can use `https://api.github.com/repos/immich-app/immich/releases/latest` to fetch the latest version.

Chiogros replied:
Yes, this one must be changed for each new release.

Thanks @panoti for your tip. 
However, looking at the latest release available could break the script when Alpine will be updated: in the future, if someone wants to install Immich on an old Alpine version (e.g. 3.14), using this link will lead the script to download an Alpine 3.15,16... Immich compatible version, which will not work.

Chiogros replied:
In top of that @alextran1502, if Dockerfiles are updated, we need to review the script to stick with the modified Dockerfile (install new packages mainly).
We don't need to change anything else while the mobile/server/web/ML code is updated.

alextran1502 commented on installation/alpine.sh:
```diff
@@ -0,0 +1,591 @@
+
+# PostgreSQL
+# Installs and initialize PostgreSQL DB.
+# Rmote database not supported.
```
Would this be as simple as setting the environment variables Immich expected for the database?

Chiogros replied:
Had forgotten about that. I do some tests and come back towards you.

alextran1502 commented on installation/alpine.sh:
```diff
@@ -0,0 +1,591 @@
+
+    # Build stage
+    echo "Installing build stage dependencies..."
+    apk add --no-cache build-base python3 libheif vips-dev ffmpeg nodejs-current npm
```
is `nodejs-current` always the `lts` version of node?

alextran1502 replied:
Maybe it is better to perform node installation with the version specified in the Dockerfile to avoid unexpected behavior of the application

Chiogros replied:
[pkgs.org](https://pkgs.org/search/?q=nodejs) gives packages versions for nodejs on different Alpine versions.
The main `nodejs` package is running node v14, where `nodejs-current` is running v16, as in Dockerfiles.
Alpine 3.14 is going [EOL](https://endoflife.date/alpine) in a few months, I don't think they will update `nodejs-current` over v16.
When Dockerfiles's Alpine version will be updated, we will have to look at the right package with a compatible `node` with Immich.

alextran1502 commented:
Thank you for the detail comments in each file. It helps a lot!

I think we want to include an installation script for Debian/Ubuntu as well. What is your plan on this?

Chiogros commented:
> What is your plan on this?

I didn't started anything for Debian/Ubuntu, but I'll start as soon as everything will be okay for Alpine (except ML).

There is one thing and a half to do to support another distro:
- find the corresponding packages for its package manager + distro + version
- for systemd family, port OpenRC services files (done once)

The longer will be the first one as you can imagine.

Commit:
Finally, remote DB works.

Commit:
1.28 compliant, npm run build in web

web/Dockerfile has been updated, changes have been mirrored in the script.

Commit:
Uninstall function

Commit:
--upgrade functionnality

Partly uninstall and delete immich components,
then compile the updated versions.

Commit:
Bugfixes, node permissions on files, executed with node.

node user now executes npm commands to set proper permissions on files
without having to do chowns again and again.
Services files now execute as the node user, a non privilegied user.

Chiogros commented:
Hi,
I finally managed how to upgrade Immich.
Everything is okay for me, I let you review what have been done since your last feedback.

panoti commented on installation/alpine.sh:
```diff
@@ -0,0 +1,781 @@
+# params: nothing
+#
+# return: nothing
+setup_redis()
```
Like database, I think we can connect to a remote Redis as well.

Commit:
Remote Redis server

Commit:
Custom DB port

Commit:
Support public login page custom message.

Commit:
Custom hostnames for non-immich services.

Commit:
Format help messages

Commit:
Optional remote nginx server.

An external nginx server can be used.
The configuration will not be applied locally, but a warning message
will invite to check the repo to see the recommended nginx config.

alextran1502 commented:
Everything looks good to me. Can you write down some words for the testing procedure and the required environment to test? Thank you for your contribution!

Commit:
Usage and testing procedures.

Give tips about script options and development procedures for future releases.

Commit:
Chown not applied on upload symlink.

A chown on a symlink without trailing / will not apply on target directory.
Also, a note has been added: the storage path should not end with a /.

Chiogros commented:
> Everything looks good to me. Can you write down some words for the testing procedure and the required environment to test? Thank you for your contribution!

Hope it was what you was looking for.
If not, please let me know!

alextran1502 commented:
@Chiogros Thank you, I will spin up an alpine VM and test then let you know the result. We probably need to move all of the instruction to the WIKI to keep things nice and clean in the readme. It is getting long :P

Chiogros commented:
Hello everybody,
I finally convinced myself to start a Docker VM.
Good luck to everyone who could want to maintain this script.
